### PR TITLE
kubeadm: Write kubelet config file to disk and persist in-cluster

### LINKED ISSUE #571: Graduate kubelet config to beta
FEATURE REQUEST

Kubelet config was introduced in `kubeadm init` as alpha feature behind feature gates, but it is planned to graduate it as soon as possible.

It should be necessary to introduce corresponding phases as well before graduation.

### LINKED ISSUE #822: Make kubeadm start controlling kubelet Config via a versioned file
## Problem statement

Currently kubeadm (the CLI) assumes a running (or actually, crashlooping) kubelet, with the _right configuration for the right kubeadm version_. This is currently achieved by packaging a kubelet dropin file together with the _kubeadm debian package_. So we have two different kubeadm entities with different responsibilities: the kubeadm CLI binary and the kubeadm deb package.

This solutions includes a number of problems:

- There is no way for kubeadm the CLI to give kubelet instructions how to run
  - Instead, we have to tell the user to go and modify the dropin manually which is time consuming and error prone
- There is no way for the cluster administrator to enforce a base kubelet configuration policy for all nodes, i.e. set the DNS domain to `cluster.global`(silly example but anyway)
- Upgrading is particularly difficult, as the kubeadm deb package bundles both the kubeadm CLI and the kubelet dropin. Why is this a problem when upgrading? Consider the following scenario:
  - You have kubeadm v1.11.0 locally, together with k8s at v1.11.0 and kubelets of v1.11.0
  - You want to upgrade to v1.12.
  - Between v1.11.x and v1.12.x, the kubelet dropin changed, as it had to. Let's say we set the `--cadvisor-port=0`flag to be secure in v1.11, but Kubernetes evolved to remove the flag in v1.12 and always turn it off. In other words, if we tried to set `--cadvisor-port=0` v1.12 the kubelet would crash.
    - This means, if we use the v1.12.x dropin for a v1.11.x kubelet we're insecure, and if we're using a v1.11.x dropin for a v1.12.x kubelet it'll crash
  - First thing, the user needs kubeadm v1.12.0 to upgrade k8s to v1.12.0, **but** the user can't do `apt-get upgrade` as that would download a newer kubelet than the API server, which is not supported (TODO reference).
  - But we can't do `apt-get install kubeadm` to only upgrade the kubeadm deb package either, as that would include a newer manifest than supported.
  - Current solution:
    - Download kubeadm the CLI only as a standalone binary using `curl`.
    - `kubeadm upgrade apply v1.12.0` => Both kubeadm and k8s version is now v1.12.0
      - Now we can upgrade the kubelet
    - `kubectl cordon $NODE && kubectl drain $NODE && apt-get upgrade`
      - Will upgrade both the kubelet and its dropin file to v1.12.
- Configuring any component with lots of knobs using CLI flags is far from optimal. Using a versioned file that supports different API versions, conversions between them and similar is far better. Hence, the community is striving towards adopting ComponentConfig for all components [reference](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit). kubeadm is already implementing ComponentConfig with its `--config` option, but our API group is at the time of writing still alpha.

#### Reference kubelet dropin created by the kubeadm deb package for v1.10.x

```
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_SYSTEM_PODS_ARGS=--pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true"
Environment="KUBELET_NETWORK_ARGS=--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin"
Environment="KUBELET_DNS_ARGS=--cluster-dns=10.96.0.10 --cluster-domain=cluster.local"
Environment="KUBELET_AUTHZ_ARGS=--authorization-mode=Webhook --client-ca-file=/etc/kubernetes/pki/ca.crt"
Environment="KUBELET_CADVISOR_ARGS=--cadvisor-port=0"
Environment="KUBELET_CERTIFICATE_ARGS=--rotate-certificates=true --cert-dir=/var/lib/kubelet/pki"
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_SYSTEM_PODS_ARGS $KUBELET_NETWORK_ARGS $KUBELET_DNS_ARGS $KUBELET_AUTHZ_ARGS $KUBELET_CADVISOR_ARGS $KUBELET_CERTIFICATE_ARGS $KUBELET_EXTRA_ARGS
```

## Proposal

Let kubeadm the CLI flow down configuration to the kubelet via a well-versioned interface. This interface will make it possible to not rely on the dropin for configuring the kubelet by neither kubeadm the CLI nor the user. Both will use the kubelet's ComponentConfiguration to configure it. kubeadm will embed the `v1beta1` ComponentConfiguration (aka `KubeletConfiguration`) of the kubelet inside of kubeadm's own Configuration file.

#### Reference kubelet dropin created by the kubeadm deb package for v1.11.x

```
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_CONFIG_ARGS=--config=/var/lib/kubelet/config.yaml"
EnvironmentFile=/etc/kubernetes/kubelet-cri.env #TBD
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_CONFIG_ARGS $KUBELET_CONTAINER_RUNTIME_ARGS $KUBELET_RESOLVER_ARGS $KUBELET_EXTRA_ARGS
```

#### kubeadm init flow

- kubeadm executes `systemctl start kubelet` if it's not already running. The kubelet starts crashlooping as it hasn't got the configuration or a KubeConfig file to talk to the API server.
- kubeadm reads `--config` and unmarshals it to a struct internally if specified. If not specified, it sets a couple of well-known defaults.
- kubeadm enforces a few security-related options for the `KubeletConfiguration` (e.g. securing the kubelet API endpoint, disabling the readonly and cAdvisor ports, etc.) in its defaulting
- kubeadm the CLI generates certificates, kubeconfig files and static Pod manifests as usual.
- kubeadm marshals the `KubeletConfiguration` and writes the bytes down to disk in the file `/var/lib/kubelet/config.yaml`.
  - This only happens if `kubelet --version` is `>= v1.11.0`, when this feature was introduced.
- The kubelet can now read its configuration and has a KubeConfig file for talking to the API server (`/etc/kubernetes/kubelet.conf`), so it can now start cleanly vs. crashlooping in systemd.
- The static Pods are started, the API server comes up, the full kubeadm `MasterConfiguration` object is stored in the `kubeadm-config` ConfigMap in the cluster, etc.
- kubeadm uploads the marshalled `KubeletConfiguration` bytes to a ConfigMap in the cluster called `kubelet-config-1.X`, where `X` is the minor version of the API Server running currently.
  - This only happens if the API Server version is `>= v1.11.0`, when this feature was introduced.
- kubeadm creates an RBAC rule that allows Bootstrap Tokens and all nodes in the cluster to read this ConfigMap.
- kubeadm continues with the rest of the init flow as usual.

#### kubeadm join flow

- kubeadm executes `systemctl start kubelet` if it's not already running. The kubelet starts crashlooping as it hasn't got the configuration or a KubeConfig file to talk to the API server.
- kubeadm contacts the API server in some way and does the discovery flow (detailed in an other proposal). The net outcome of this process is a KubeConfig client to use for the TLS bootstrap. This KubeConfig object is written down to `/etc/kubernetes/kubelet-bootstrap.conf`.
- In the usual flow the authentication method for the TLS bootstrap is using a Bootstrap Token. Thanks to creating that RBAC rule earlier, kubeadm uses the Bootstrap Token to download the desired `KubeletConfiguration` to `/var/lib/kubelet/config.yaml`.
  - This only happens if `kubelet --version` is `>= v1.11.0`, when this feature was introduced.
- The kubelet can now read its configuration and has a bootstrap KubeConfig file for talking to the API server (`/etc/kubernetes/bootstrap-kubelet.conf`), so it can now start cleanly vs. crashlooping in systemd.
  - The kubelet performs the TLS bootstrap which yields it an unique clientcert credential stored in `/etc/kubernetes/kubelet.conf`.
- The kubelet can now function normally.

#### v1.10 -> v1.11 upgrades

- The user created a kubeadm cluster with kubeadm, kubelet and the API server at `v1.10.2`, but now wants to upgrade to `v1.11.0`. The `v1.10.2` kubeadm kubelet dropin exists locally.
- The dropin for `v1.11.0` is different than `v1.10.2` due to that now it references `--config=/var/lib/kubelet/config.yaml` 
  - Worth noting is that it would be different in any case, as we would have made changes to the dropin in any case, e.g. disabling the readonly port, enabling Token authentication to the kubelet API, etc.
- We need to do the same flow as earlier, to download the kubeadm CLI independently and upgrade the API server first, and after that cordon, drain, and `apt-get upgrade` in order to upgrade the kubelet and get the new `v1.11.0` dropin at the right time.
- What does `kubeadm upgrade apply v1.11.0` do?
  - It first executes a set of preflight checks, etc. and then upgrades the control plane components in order
  - It updates the `kubeadm-config` ConfigMap.
  - It writes a new `kubelet-config-1.11` ConfigMap with the desired configuration for `v1.11.x` kubelets.
  - It also writes these v1.11-specific marshalled `KubeletConfiguration` bytes down to disk in the `/var/lib/kubelet/config.yaml` file, so the kubelet later can pick up the desired configuration when upgraded.
  - Then the master node is upgraded successfully.
- In order to upgrade the nodes, the following should be done:
  - Cordon/drain from the master node as usual.
  - `kubeadm alpha phase kubelet write-config-to-disk` to fetch the desired v1.11 configuration down to disk.
  - `apt-get upgrade` which upgrades the kubeadm CLI binary and the kubelet itself

#### v1.11 -> upgrades

- As the kubeadm-kubelet dropin file doesn't change anymore, we can skip the `curl`-ing down of the standalone kubeadm binary in the master upgrade flow, and instead tell the user to run `apt-get install kubeadm=1.11.0-00`
- We don't have to worry about the kubeadm deb messing things up for the kubelet deb anymore.

### Open questions

- Should we embed the `KubeletConfiguration` in the `kubeadm-config` ConfigMap as well, or set it to `nil` before uploading as it's stored
- What should the command to upgrade the kubelet? `kubeadm upgrade node config`?

PR to implement this is available to look at https://github.com/kubernetes/kubernetes/pull/63887

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
In order to make configuration flow from the cluster level to node level, we need a way for kubeadm to tell the kubelet what config to use. As of v1.10 (I think) the kubelet can read `--config` using the kubelet Beta ComponentConfiguration API, so now we have an interface to talk to the kubelet properly.

This PR:
 - Writes the kubelet ComponentConfig to `/var/lib/kubelet/config.yaml` on init and join
 - Writes an environment file to source in the kubelet systemd dropin `/var/lib/kubelet/kubeadm-flags.env`. This file contain runtime flags that should be passed to the kubelet.
 - Uploads a ConfigMap with the name `kubelet-config-1.X`
 - Patches the node object so that it starts using the ConfigMap with updates using Dynamic Kubelet Configuration, **only if the feature gate is set** (currently alpha and off by default, not intended to be switched on in v1.11)
 - Updates the phase commands to reflect this new flow

The kubelet dropin file I used now looks like this:
```
# v1.11.x dropin as-is at HEAD
# /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
---
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_CONFIG_ARGS=--config=/var/lib/kubelet/config.yaml"
EnvironmentFile-=/var/lib/kubelet/kubeadm-flags.env
# Should default to 0 in v1.11: https://github.com/kubernetes/kubernetes/pull/63881, and hence not be here in the real v1.11 manifest
Environment="KUBELET_CADVISOR_ARGS=--cadvisor-port=0"
# Should be configurable via the config file: https://github.com/kubernetes/kubernetes/issues/63878, and hence be configured using the file in v1.11
Environment="KUBELET_CERTIFICATE_ARGS=--rotate-certificates=true"
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_CONFIG_ARGS $KUBELET_KUBEADM_ARGS $KUBELET_CADVISOR_ARGS $KUBELET_CERTIFICATE_ARGS $KUBELET_EXTRA_ARGS
---
# v1.11.x dropin end goal
# /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
---
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_CONFIG_ARGS=--config=/var/lib/kubelet/config.yaml"
EnvironmentFile-=/var/lib/kubelet/kubeadm-flags.env
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_CONFIG_ARGS $KUBELET_KUBEADM_ARGS $KUBELET_EXTRA_ARGS
---
# Environment file dynamically created at runtime by "kubeadm init"
# /var/lib/kubelet/kubeadm-flags.env
KUBELET_KUBEADM_ARGS=--cni-bin-dir=/opt/cni/bin --cni-conf-dir=/etc/cni/net.d --network-plugin=cni
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/822
Fixes https://github.com/kubernetes/kubeadm/issues/571

**Special notes for your reviewer**:

**Release note**:

```release-note
"kubeadm init" now writes a structured and versioned kubelet ComponentConfiguration file to `/var/lib/kubelet/config.yaml` and an environment file with runtime flags (you can source this file in the systemd kubelet dropin) to `/var/lib/kubelet/kubeadm-flags.env`.
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @mtaufen

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -425,6 +411,19 @@ func (i *Init) Run(out io.Writer) error {
 		return fmt.Errorf("error uploading configuration: %v", err)
 	}
 
+	glog.V(1).Infof("[init] creating kubelet configuration configmap")
+	if err := kubeletphase.CreateKubeletConfigMap(i.cfg, client); err != nil {
+		return fmt.Errorf("error creating base kubelet configuration: %v", err)
+	}
+
+	// NOTE: flag "--dynamic-config-dir" should be specified in /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: b/c the kubelet will write to this location it should be /var/lib/kubelet ...

Comment (timothysc) [reply cmd/kubeadm/app/cmd/init.go]: How are you going to upgrade into this?  You will need a systemctl daemon-reload + restart.

File: cmd/kubeadm/app/phases/kubelet/config.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+				Kind: rbac.GroupKind,
+				Name: kubeadmconstants.NodeBootstrapTokenAuthGroup,
+			},
+		},
+	})
+}
+
+// DownloadKubeletConfig downloads the kubelet configuration from a ConfigMap and writes it to disk.
+// Used at "kubeadm join" time
+func DownloadKubeletConfig(kubeletKubeConfig string) error {
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/kubelet/config.go]: shouldn't we have a download-config subcommand as well then for debugging?

**Review by timothysc — COMMENTED** [review]
So in general this looks like an overall cleanup improvement.. But upgrading from an old unit file seems fraught with peril 

/cc @detiber

File: cmd/kubeadm/app/constants/constants.go [hunk 3]
```diff
@@ -191,17 +188,17 @@ const (
...
 
-	// KubeletBaseConfigurationDir specifies the directory on the node where stores the initial remote configuration of kubelet
-	KubeletBaseConfigurationDir = "/var/lib/kubelet/config/init"
+	// KubeletBaseConfigMapRolePrefix defines the base kubelet configuration ConfigMap.
+	KubeletBaseConfigMapRolePrefix = "kubeadm:kubelet-config-"
 
-	// KubeletBaseConfigurationFile specifies the file name on the node which stores initial remote configuration of kubelet
-	KubeletBaseConfigurationFile = "kubelet"
+	// KubeletConfigurationFile specifies the file name on the node which stores initial remote configuration of kubelet
+	KubeletConfigurationFile = "/var/lib/kubelet/config.yaml"
```
Comment (detiber) [inline cmd/kubeadm/app/constants/constants.go]: If I am understanding the docs for [Dynamic Kubelet Config](https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/), it looks like the dynamic config is downloaded from the config map to the directory specified by `--dynamic-config-dir` automatically and the config pulled down overrides the values that are set through `--config`. With that being the case, shouldn't we write the static config file (or initial config in the case of dynamic config) to `/etc/kubernetes` and leave `/var/lib/kubelet` for the content that is dynamically managed at runtime by the kubelet?

**Comment by detiber** [comment]
> The kubelet dropin file I used now looks like this:
>  
> ```
> # v1.11.x dropin
> # /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
> ---
> [Service]
> Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
> Environment="KUBELET_CONFIG_ARGS=--config=/var/lib/kubelet/config.yaml"
> EnvironmentFile=/etc/kubernetes/kubelet-cri.env
> # Should default to 0 in v1.11: https://github.com/kubernetes/kubernetes/pull/63881, and hence not be here in the real v1.11 manifest
> Environment="KUBELET_CADVISOR_ARGS=--cadvisor-port=0"
> # Should be configurable via the config file: https://github.com/kubernetes/kubernetes/issues/63878, and hence be configured using the file in v1.11
> Environment="KUBELET_CERTIFICATE_ARGS=--rotate-certificates=true"
> ExecStart=
> ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_CONFIG_ARGS $KUBELET_CONTAINER_RUNTIME_ARGS $KUBELET_CADVISOR_ARGS $KUBELET_CERTIFICATE_ARGS $KUBELET_RESOLVER_ARGS $KUBELET_EXTRA_ARGS
> ---
> # /etc/kubernetes/kubelet-cri.env
> # In this file, you can easily add support for other CRI runtimes and specify flags needed for that CRI
> # By default these arguments are here as they are specific to the docker runtime
> KUBELET_CONTAINER_RUNTIME_ARGS=--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin
> ```

The addition of an environment file is great here, but if we go forward with this it would probably be a good idea to stick with standard env file paths for the OSs we ship packages for. I know for CentOS/Fedora this would be `/etc/sysconfig/kubelet` and `/etc/default/kubelet` for Ubuntu

An aside while I'm on the topic of systemd unit configuration managed  by the packaging, we really shouldn't be placing packaged managed unit files and drop-in files in `/etc/systemd/system`, instead they should be placed in `/usr/lib/systemd/system`

I'm still concerned with how we handle this transition during upgrades, since there is a bit of chicken/egg involved between the package managed dropin and the configuration done by kubeadm upgrade. If we force the update with the package upgrade of kubeadm, then we risk breaking the running kubelet prior to install. However, we *could* transition the ownership of the drop-in file to kubeadm, which would allow for kubeadm itself to mutate the drop-in, perform a `systemctl daemon-reload` and then restart the kubelet. This would also give us a way to backup the drop-in and revert it if needed as well.

**Comment by randomvariable** [comment]
Agree that we should follow the distribution standards for placement for environment files and using `/usr/lib/systemd` over `/etc/systemd/system`

There's a potential issue at present in that the docs ask people to modify `10-kubeadm.conf` to set the cgroup driver, and yet we haven't also marked these files as conffiles in Bazel, so package updates will blow the user changes away and would probably break their system if the cgroup driver is non-default?

Moving management of that particular config to something kubeadm controls so that the drop-ins are truly static makes more sense.

**Comment by neolit123** [comment]
@randomvariable 

> There's a potential issue at present in that the docs ask people to modify 10-kubeadm.conf to set the cgroup driver, and yet we haven't also marked these files as conffiles in Bazel, so package updates will blow the user changes away and would probably break their system if the cgroup driver is non-default?

the already modified kubelet config by the user (e.g. cgroup driver updated) is a problem. i think the documentation and the suggested workflow for the kubelet would have been to never modify the kublet config but handle extra arguments from the environment only. if the same location is written to, a backup of the user config should be created - e.g. `$cp 10-kubeadm.conf 10-kubeadm.conf.user` every time.

@detiber 

> I'm still concerned with how we handle this transition during upgrades, since there is a bit of chicken/egg involved between the package managed dropin and the configuration done by kubeadm upgrade. If we force the update with the package upgrade of kubeadm, then we risk breaking the running kubelet prior to install. However, we could transition the ownership of the drop-in file to kubeadm, which would allow for kubeadm itself to mutate the drop-in, perform a systemctl daemon-reload and then restart the kubelet. This would also give us a way to backup the drop-in and revert it if needed as well.

do you have something in mind on how the mutator would work? if the user has modified his kubelet config a lot, doing a `diff -u <old> <new> > patchfile; patch < patchfile` might not be sufficient. perhaps a line by line parser / comparator of sorts is needed here?

**Comment by detiber** [comment]
One option that we have is to have the upgrade scripts in packaging preserve the existing file and create the new version of the file under /usr/lib/systemd and require the user to reconcile the differences. This is a more manual step, but would get us to the place that we want to be...  We could also tell users that if they haven't modified the file that it is safe to just blow it away.

As for documentation on how to modify the unit config, we could tell users to just use `systemctl edit kubelet` which will open an editor and create a drop-in file under /etc/systemd/system/kubelet.service.d for the user.

**Comment by neolit123** [comment]
going back to this:

> An aside while I'm on the topic of systemd unit configuration managed by the packaging, we really shouldn't be placing packaged managed unit files and drop-in files in /etc/systemd/system, instead they should be placed in /usr/lib/systemd/system

this is so true. `usr/lib` is the de-facto vendor specific location for packaged software.

> One option that we have is to have the upgrade scripts in packaging preserve the existing file and create the new version of the file under /usr/lib/systemd and require the user to reconcile the differences. This is a more manual step, but would get us to the place that we want to be... We could also tell users that if they haven't modified the file that it is safe to just blow it away.

i think the files should be placed in the same folder and for the old one it should a suffix should be appended, like `.user` or `.old` . having the old / new in different locations is even more confusing.

> As for documentation on how to modify the unit config, we could tell users to just use systemctl edit kubelet which will open an editor and create a drop-in file under /etc/systemd/system/kubelet.service.d for the user.

this seems like a good solution.

so in terms of the docs and `website` i'm seeing a bad trend where a lot of the reports are about the cgroup driver issue. ideally the user should never edit system service configuration files as part of the UX...ever.
yet i don't see this being possible unless things like `--cgroup-driver=auto` are the default option.

**Comment by detiber** [comment]
> > One option that we have is to have the upgrade scripts in packaging preserve the existing file and create the new version of the file under /usr/lib/systemd and require the user to reconcile the differences. This is a more manual step, but would get us to the place that we want to be... We could also tell users that if they haven't modified the file that it is safe to just blow it away.

> 
> i think the files should be placed in the same folder and for the old one a suffix should be appended, like .user or .old . having the old / new in different locations is even more confusing.

I do agree that it is confusing, but it would ensure that the existing drop-in would override the package provided one...  We could always attempt to do a diff against the known previous version and only leave the diff as a drop-in under /etc/systemd/system/kubelet.service.d. We could also provide a preflight check to force the user to take some action prior to upgrade as well.

I think anything we do for the transition is going to be awkward, but as long as we provide sufficient documentation and tooling to assist the user through the transition then we can limit the impact.

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 4]
```diff
@@ -18,184 +18,161 @@ package phases
...
-	kubeadmscheme.Scheme.Default(cfg)
+func NewCmdKubeletEnableDynamicConfig(kubeConfigFile *string) *cobra.Command {
+	var nodeName, kubeletVersionStr string
 
-	var cfgPath string
 	cmd := &cobra.Command{
-		Use:     "enable",
-		Aliases: []string{"update"},
-		Short:   "Enables or updates dynamic kubelet configuration on node",
+		Use:     "enable-dynamic-config",
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/kubelet.go]: This is going to be off by default, right?  I don't see a message stating the default.

File: cmd/kubeadm/app/phases/kubelet/config.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+	rbachelper "k8s.io/kubernetes/pkg/apis/rbac/v1"
+	kubeletconfigscheme "k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/scheme"
+	kubeletconfigv1beta1 "k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/v1beta1"
+	"k8s.io/kubernetes/pkg/util/version"
+)
+
+// WriteKubeletConfigToDisk writes the kubelet config object down to a file
+// Used at "kubeadm init" and "kubeadm upgrade" time
+func WriteKubeletConfigToDisk(kubeletConfig *kubeletconfigv1beta1.KubeletConfiguration, kubeletVersion *version.Version) error {
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/kubelet/config.go]: b/c this is in a "kubelet" package we can nix the "Kubelet" bitz.

File: cmd/kubeadm/app/phases/kubelet/dynamic.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+// Used at "kubeadm init", "kubeadm join" and "kubeadm upgrade" time
+func EnableDynamicConfigForNode(client clientset.Interface, nodeName string, kubeletVersion *version.Version) error {
+
+	// If the kubelet version is v1.10.x, exit
+	if kubeletVersion.LessThan(kubeadmconstants.MinimumKubeletConfigVersion) {
+		return nil
+	}
+
+	configMapName := kubeletConfigMapName(kubeletVersion)
+	fmt.Printf("[kubelet] Enabling Dynamic Kubelet Config for Node %q; config sourced from ConfigMap %q in namespace %s\n",
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/kubelet/dynamic.go]: Shouldn't we put warnings about the state of this feature and that folks can easily brick a whole cluster?

File: cmd/kubeadm/app/phases/kubelet/flags.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+)
+
+const kubeletEnvFile = "/var/lib/kubelet/kubeadm-flags.env"
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/kubelet/flags.go]: This should probably be by the other vars.

**Review by timothysc — CHANGES_REQUESTED** [review]
I think we can/should split this up.  

There are 3 things going on.  
1. Kubelet Config Changes 
2. Main Config Changes 
3. DynamicKubelet Config which should definitely be seaprate imo.

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 8]
```diff
@@ -18,184 +18,161 @@ package phases
...
-			// This call returns the ready-to-use configuration based on the configuration file that might or might not exist and the default cfg populated by flags
-			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
+			// This call returns the ready-to-use configuration based on the configuration file
+			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, &kubeadmapiv1alpha2.MasterConfiguration{})
+			kubeadmutil.CheckErr(err)
+
+			client, err := kubeconfigutil.ClientSetFromFile(*kubeConfigFile)
+			kubeadmutil.CheckErr(err)
+
+			err = kubeletphase.CreateKubeletConfigMap(internalcfg, client)
```
[the remark this answers is not in the document]
Comment (luxas) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: done

File: cmd/kubeadm/app/phases/kubelet/config.go (hunk 5 above)
Comment (luxas) [reply cmd/kubeadm/app/phases/kubelet/config.go]: done

File: cmd/kubeadm/app/phases/kubelet/flags.go (hunk 7 above)
Comment (luxas) [reply cmd/kubeadm/app/phases/kubelet/flags.go]: done

File: cmd/kubeadm/app/cmd/phases/kubelet.go (hunk 4 above)
Comment (luxas) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: yes, off by default. Now added a scary warning

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

Next time please break-up the PRs according to finite chunks of behavior, this PR is indeed 3 separate things.

**Comment by luxas** [comment]
goh forgot to regenerate bazel

> Next time please break-up the PRs according to finite chunks of behavior, this PR is indeed 3 separate things.

Normally I do that, but this is to a large extent a refactor of what's already there, so I thought it was better to rewrite what's there in one step to conform to the guidelines we now have vs rewrite the exact same thing in smaller chunks. However, thanks for the review!

**Pull request merged**